File: backend/src/analysis/predictor.py
import pandas as pd
import numpy as np
from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class PredictedHour:
    timestamp: str
    price_eur_mwh: float
    price_huf_mwh: float
    is_cheap: bool
    is_expensive: bool
    is_forecast: bool
    ci_lower_huf_mwh: float = 0.0
    ci_upper_huf_mwh: float = 0.0


def predict_prices(historical: pd.Series, days_ahead: int = 7) -> pd.Series:
    """
    Előre jelzi az árakat az elmúlt időszak mintázata alapján.

    Modell:
      - (hét napja, óra) szerinti átlagos árakon alapul
      - az elmúlt 7 nap trendjével korrigálva
      - kis véletlenszerű zajjal a realisztikusabb megjelenés érdekében
    """
    if historical.empty:
        return pd.Series(dtype=float, name="price_eur_mwh")

    df = historical.to_frame("price")
    df["hour"] = df.index.hour
    df["dayofweek"] = df.index.dayofweek

    # Alap minta: (hét napja, óra) szerinti átlag
    pattern = df.groupby(["dayofweek", "hour"])["price"].mean()

    # Trend: elmúlt 7 nap vs. teljes időszak átlag
    cutoff = historical.index[-1] - pd.Timedelta(days=7)
    recent = historical[historical.index >= cutoff]
    overall_mean = float(historical.mean())
    recent_mean = float(recent.mean()) if not recent.empty else overall_mean
    trend_factor = recent_mean / overall_mean if overall_mean != 0 else 1.0

    # Szórás a historikus adatokból
    hourly_std = df.groupby(["dayofweek", "hour"])["price"].std().fillna(5.0)

    # Előrejelzési idősor: holnaptól számítva
    tz = historical.index.tz
    start = (historical.index[-1] + pd.Timedelta(hours=1)).floor("h")
    future_index = pd.date_range(start=start, periods=days_ahead * 24, freq="1h", tz=tz)

    np.random.seed(42)
    predictions = []
    for ts in future_index:
        dow = ts.dayofweek
        hour = ts.hour
        base = pattern.get((dow, hour), overall_mean)
        std = float(hourly_std.get((dow, hour), 5.0))
        noise = np.random.normal(0, std * 0.3)
        pred = base * trend_factor + noise
        predictions.append(float(pred))

    return pd.Series(predictions, index=future_index, name="price_eur_mwh")


def compute_confidence_intervals(
    historical: pd.Series,
    forecast: pd.Series,
    sigma: float = 1.0,
) -> tuple:
    """
    Konfidencia intervallum az előrejelzett pontokhoz a historikus szórás alapján.

    Args:
        historical: Historikus árak
        forecast: Előrejelzett árak (pd.Series)
        sigma: Szórás szorzó (1.0 ≈ 68%-os CI, 1.96 ≈ 95%-os CI)

    Returns:
        (ci_lower, ci_upper) tuple of pd.Series
    """
    df = historical.to_frame("price")
    df["hour"] = df.index.hour
    df["dayofweek"] = df.index.dayofweek
    hourly_std = df.groupby(["dayofweek", "hour"])["price"].std().fillna(5.0)

    ci_lowers = []
    ci_uppers = []
    for ts, price in forecast.items():
        dow = ts.dayofweek
        hour = ts.hour
        std = float(hourly_std.get((dow, hour), 5.0))
        ci_lowers.append(max(0.0, price - sigma * std))
        ci_uppers.append(price + sigma * std)

    return (
        pd.Series(ci_lowers, index=forecast.index),
        pd.Series(ci_uppers, index=forecast.index),
    )


def compute_model_accuracy(historical: pd.Series) -> dict:
    """
    Becsüli a modell pontosságát walk-forward validációval.
    Az utolsó 7 nap adatait az azt megelőző időszak alapján jelzi előre,
    majd összeveti a valódi értékekkel.

    Returns:
        dict: mape_pct (MAPE %) és validation_days
    """
    MIN_DAYS = 14
    if len(historical) < MIN_DAYS * 24:
        # Nincs elég adat a validációhoz — CV alapú becslés
        mean = float(historical.mean())
        std = float(historical.std())
        cv_pct = round((std / mean) * 100, 1) if mean > 0 else 20.0
        return {"mape_pct": cv_pct, "validation_days": 0, "method": "cv"}

    # Train: minden adat az utolsó 7 nap előtt, Test: utolsó 7 nap
    cutoff = historical.index[-1] - pd.Timedelta(days=7)
    train = historical[historical.index < cutoff]
    test = historical[historical.index >= cutoff]

    if train.empty or test.empty:
        return {"mape_pct": 20.0, "validation_days": 0, "method": "fallback"}

    # Előrejelzés a train alapján
    pred = predict_prices(train, days_ahead=8)

    # Közös időpontok
    common_idx = test.index.intersection(pred.index)
    if len(common_idx) == 0:
        return {"mape_pct": 20.0, "validation_days": 0, "method": "fallback"}

    actual = test[common_idx].values
    predicted = pred[common_idx].values

    # MAPE kiszámítása (nullával való osztás elkerülése)
    mask = np.abs(actual) > 1.0
    if not mask.any():
        return {"mape_pct": 20.0, "validation_days": 7, "method": "walkforward"}

    mape = float(np.mean(np.abs((actual[mask] - predicted[mask]) / actual[mask])) * 100)
    return {
        "mape_pct": round(min(mape, 99.0), 1),
        "validation_days": 7,
        "method": "walkforward",
    }


def combined_forecast(
    historical: pd.Series,
    forecast: pd.Series,
    eur_huf: float,
    days_ahead: int = 7,
    history_days: int = 7,
) -> List[PredictedHour]:
    """Összekapcsolt historikus + előrejelzett adatsor a frontendnek, konfidencia intervalumokkal."""

    CHEAP_Q = 0.25
    EXP_Q = 0.75

    combined = pd.concat([historical, forecast])
    q_low = float(combined.quantile(CHEAP_Q))
    q_high = float(combined.quantile(EXP_Q))

    # Konfidencia intervallumok az előrejelzett adatokra
    ci_lower, ci_upper = compute_confidence_intervals(historical, forecast, sigma=1.0)

    result = []

    # Historikus adatok (utolsó history_days nap)
    cutoff = historical.index[-1] - pd.Timedelta(days=history_days)
    hist_slice = historical[historical.index >= cutoff]
    for ts, price in hist_slice.items():
        huf = round(float(price) * eur_huf, 0)
        result.append(PredictedHour(
            timestamp=ts.isoformat(),
            price_eur_mwh=round(float(price), 2),
            price_huf_mwh=huf,
            is_cheap=bool(float(price) <= q_low),
            is_expensive=bool(float(price) >= q_high),
            is_forecast=False,
            ci_lower_huf_mwh=huf,
            ci_upper_huf_mwh=huf,
        ))

    # Előrejelzett adatok CI-vel
    for ts, price in forecast.items():
        cl = float(ci_lower.get(ts, price))
        cu = float(ci_upper.get(ts, price))
        result.append(PredictedHour(
            timestamp=ts.isoformat(),
            price_eur_mwh=round(float(price), 2),
            price_huf_mwh=round(float(price) * eur_huf, 0),
            is_cheap=bool(float(price) <= q_low),
            is_expensive=bool(float(price) >= q_high),
            is_forecast=True,
            ci_lower_huf_mwh=round(cl * eur_huf, 0),
            ci_upper_huf_mwh=round(cu * eur_huf, 0),
        ))

    return result

File: mobile/src/domain/predictor.ts
import type { HourPoint, ModelAccuracy, PricePoint } from './types';
import { zonedParts } from './time';

const HOUR = 3600000;
const DAY = 24 * HOUR;

// Determinisztikus véletlengenerátor (a Python np.random.seed(42) megfelelője célra:
// azonos bemenetre azonos előrejelzés).
function mulberry32(seed: number) {
  let a = seed >>> 0;
  return () => {
    a = (a + 0x6d2b79f5) >>> 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

function gaussian(rng: () => number): number {
  let u = 0;
  let v = 0;
  while (u === 0) u = rng();
  while (v === 0) v = rng();
  return Math.sqrt(-2.0 * Math.log(u)) * Math.cos(2.0 * Math.PI * v);
}

/** numpy/pandas-féle lineáris interpolációs kvantilis. */
export function quantile(values: number[], q: number): number {
  if (!values.length) return NaN;
  const s = [...values].sort((a, b) => a - b);
  const pos = (s.length - 1) * q;
  const lo = Math.floor(pos);
  const hi = Math.ceil(pos);
  if (lo === hi) return s[lo];
  return s[lo] + (s[hi] - s[lo]) * (pos - lo);
}

function mean(xs: number[]): number {
  return xs.length ? xs.reduce((a, b) => a + b, 0) / xs.length : 0;
}

/** Minta-szórás (ddof = 1), mint a pandas std(). 1 elemnél NaN. */
function sampleStd(xs: number[]): number {
  if (xs.length < 2) return NaN;
  const m = mean(xs);
  return Math.sqrt(xs.reduce((a, b) => a + (b - m) ** 2, 0) / (xs.length - 1));
}

interface Pattern {
  meanBy: Map<number, number>;
  stdBy: Map<number, number>;
}

const groupKey = (t: number) => {
  const p = zonedParts(new Date(t));
  return p.dow * 24 + p.hour;
};

function buildPattern(hist: HourPoint[]): Pattern {
  const groups = new Map<number, number[]>();
  for (const h of hist) {
    const k = groupKey(h.t);
    const g = groups.get(k);
    if (g) g.push(h.p);
    else groups.set(k, [h.p]);
  }
  const meanBy = new Map<number, number>();
  const stdBy = new Map<number, number>();
  for (const [k, xs] of groups) {
    meanBy.set(k, mean(xs));
    const s = sampleStd(xs);
    stdBy.set(k, Number.isNaN(s) ? 5.0 : s);
  }
  return { meanBy, stdBy };
}

/**
 * Árelőrejelzés: (hét napja, óra) szerinti átlag × az elmúlt 7 nap trendje + kis zaj.
 * A historikus sor utolsó órája utáni órától indul.
 */
export function predictPrices(hist: HourPoint[], daysAhead = 7): HourPoint[] {
  if (!hist.length) return [];
  const sorted = [...hist].sort((a, b) => a.t - b.t);
  const { meanBy, stdBy } = buildPattern(sorted);
  const last = sorted[sorted.length - 1].t;
  const cutoff = last - 7 * DAY;
  const recent = sorted.filter((h) => h.t >= cutoff).map((h) => h.p);
  const overallMean = mean(sorted.map((h) => h.p));
  const recentMean = recent.length ? mean(recent) : overallMean;
  const trend = overallMean !== 0 ? recentMean / overallMean : 1.0;

  const rng = mulberry32(42);
  const start = Math.floor((last + HOUR) / HOUR) * HOUR;
  const out: HourPoint[] = [];
  for (let i = 0; i < daysAhead * 24; i++) {
    const t = start + i * HOUR;
    const k = groupKey(t);
    const base = meanBy.get(k) ?? overallMean;
    const std = stdBy.get(k) ?? 5.0;
    const noise = gaussian(rng) * std * 0.3;
    out.push({ t, p: base * trend + noise });
  }
  return out;
}

/** Konfidencia-sáv az előrejelzett pontokhoz a historikus szórás alapján. */
export function confidenceIntervals(hist: HourPoint[], forecast: HourPoint[], sigma = 1.0) {
  const { stdBy } = buildPattern(hist);
  return forecast.map((f) => {
    const std = stdBy.get(groupKey(f.t)) ?? 5.0;
    return { t: f.t, lower: Math.max(0, f.p - sigma * std), upper: f.p + sigma * std };
  });
}

/** Walk-forward validáció: az utolsó 7 nap a megelőző időszakból előrejelezve (MAPE %). */
export function computeModelAccuracy(hist: HourPoint[]): ModelAccuracy {
  const MIN_DAYS = 14;
  const ps = hist.map((h) => h.p);
  if (hist.length < MIN_DAYS * 24) {
    const m = mean(ps);
    const s = sampleStd(ps);
    const cv = m > 0 && !Number.isNaN(s) ? Math.round((s / m) * 1000) / 10 : 20.0;
    return { mape_pct: cv, validation_days: 0, method: 'cv' };
  }
  const sorted = [...hist].sort((a, b) => a.t - b.t);
  const cutoff = sorted[sorted.length - 1].t - 7 * DAY;
  const train = sorted.filter((h) => h.t < cutoff);
  const test = sorted.filter((h) => h.t >= cutoff);
  if (!train.length || !test.length) return { mape_pct: 20.0, validation_days: 0, method: 'fallback' };
  const pred = new Map(predictPrices(train, 8).map((h) => [h.t, h.p]));
  const errs: number[] = [];
  for (const h of test) {
    const p = pred.get(h.t);
    if (p == null || Math.abs(h.p) <= 1.0) continue;
    errs.push(Math.abs((h.p - p) / h.p));
  }
  if (!errs.length) return { mape_pct: 20.0, validation_days: 7, method: 'walkforward' };
  const mape = mean(errs) * 100;
  return { mape_pct: Math.round(Math.min(mape, 99.0) * 10) / 10, validation_days: 7, method: 'walkforward' };
}

const r2 = (x: number) => Math.round(x * 100) / 100;

/** Historikus (utolsó history_days nap) + előrejelzett órák egy listában, olcsó/drága jelöléssel. */
export function combinedForecast(hist: HourPoint[], forecast: HourPoint[], eurHuf: number, historyDays = 7): PricePoint[] {
  if (!hist.length) return [];
  const sorted = [...hist].sort((a, b) => a.t - b.t);
  const all = [...sorted.map((h) => h.p), ...forecast.map((h) => h.p)];
  const qLow = quantile(all, 0.25);
  const qHigh = quantile(all, 0.75);
  const ci = new Map(confidenceIntervals(sorted, forecast).map((c) => [c.t, c]));
  const cutoff = sorted[sorted.length - 1].t - historyDays * DAY;
  const out: PricePoint[] = [];
  for (const h of sorted) {
    if (h.t < cutoff) continue;
    const hufMwh = Math.round(h.p * eurHuf);
    out.push({
      timestamp: new Date(h.t).toISOString(),
      price_eur_mwh: r2(h.p),
      price_huf_mwh: hufMwh,
      price_huf_kwh: r2(hufMwh / 1000),
      is_cheap: h.p <= qLow,
      is_expensive: h.p >= qHigh,
      is_forecast: false,
      ci_lower_huf_kwh: r2(hufMwh / 1000),
      ci_upper_huf_kwh: r2(hufMwh / 1000),
    });
  }
  for (const f of forecast) {
    const c = ci.get(f.t);
    const hufMwh = Math.round(f.p * eurHuf);
    out.push({
      timestamp: new Date(f.t).toISOString(),
      price_eur_mwh: r2(f.p),
      price_huf_mwh: hufMwh,
      price_huf_kwh: r2(hufMwh / 1000),
      is_cheap: f.p <= qLow,
      is_expensive: f.p >= qHigh,
      is_forecast: true,
      ci_lower_huf_kwh: r2(Math.round((c?.lower ?? f.p) * eurHuf) / 1000),
      ci_upper_huf_kwh: r2(Math.round((c?.upper ?? f.p) * eurHuf) / 1000),
    });
  }
  return out;
}
